src/coreclr/src/debug/dbgutil/elfreader.cpp

```diff
@@ -172,6 +170,50 @@ ElfReader::PopulateForSymbolLookup(uint64_t baseAddress)
         return false;
     }
 
+    uint64_t minAddr = std::min<uint64_t>(std::min<uint64_t>((uint64_t)m_gnuHashTableAddr,
+                                                             (uint64_t)m_stringTableAddr),
+                                                             (uint64_t)m_symbolTableAddr);
+    uint64_t maxAddr = std::max<uint64_t>(std::max<uint64_t>((uint64_t)m_gnuHashTableAddr + sizeof(GnuHashTable) + sizeof(int32_t),
+                                                             (uint64_t)m_stringTableAddr + m_stringTableSize),
+                                                             (uint64_t)m_symbolTableAddr);
+
+    bool addLoadBias = false;
+
+    if (minAddr < baseAddress)
+    {
+        // All the dynamic table pointers must point inside the load region of the module
+        // Therefore if any are less than the baseAddress we must add the loadbias.
+        addLoadBias = true;
+    }
+    else if (maxAddr > loadsize)
+    {
+        // All the dynamic table pointers must point inside the load region of the module
+        // Since the maxAddr is greater than the loadsize we can conclusively
+        // determine that we do not need to add the loadbias
+        // This is because maxAddr + baseAddr would be outside the module (greater than loadsize + baseAddress).
+        addLoadBias = false;
+    }
+    else
+    {
+        // We cannot conclusively determine whether we need to add the load bias
```

Comment:
I am still leaning to this approach.

+ This case is theoretically possible, but difficult to test because it has not occurred. 
+ I believe the hardening of the symbol reader, `ReadElfMemory`, makes this code reasonably safe. 
+ `ReadElfMemory` is also not terribly complicated.
+ `ReadElfMemory` also may provide other benefits in cases of corrupt dumps or elf files.

There are alternatives.
+ We could simply `return false;` in this case.  Revert unnecessary code.
+ Ignore the possibility of this case. Drop lines 176-178 & 188-208. Drop other unnecessary code.
+ Fallback to an `#ifdef LINUX_ALPINE` for this case.
+ Abandon the possibility of dynamically supporting Linux & Alpine.
+ Define a mechanism for the hosting layer to inform the elf-reader the type of dump.

Reply:
In preference order, these are the leading options

1. This approach.
2. Simply `return false;`
3. Abandon the possibility of dynamically supporting Linux & Alpine.

Reply:
> * I believe the hardening of the symbol reader, `ReadElfMemory`, makes this code reasonably safe.

If the ranges are correct, yeah. The only way that if could fail would be a symbol falls in an address in a PT_LOAD gap. This seems unlikely.

> * `ReadElfMemory` is also not terribly complicated.
> * `ReadElfMemory` also may provide other benefits in cases of corrupt dumps or elf files.

There's not that many cases I can think of where we'd get a dump corrupted in such a way that we'd end reading in one of those regions (we'd need a corrupt .symtab, or a missing segment). That being said `ReadElfMemory` doesn't feel like the complicated piece. However, it still relies on the stored state that Mike said (I still don't know why. A comment might be warranted...)?

The other part I like of this change is not calling populate and then trying to get the symbol. Felt a little of - remember to do this. Tends to be a step that's easy to forget. 

> * We could simply `return false;` in this case.  Revert unnecessary code.
> * Ignore the possibility of this case. Drop lines 176-178 & 188-208. Drop other unnecessary code.

I am still not sure 176-178 are correct, so I am not sure the condition on 188 (the comment in that branch should be `(greater than loadsize + loadbias)` given that you are using the vaddr in the calculations). I expect 182-187 to be the heavily predominant branch on musl based distros, while the second one should be the base for glibc based distros (not sure what about bionic's libc)

> * Fallback to an `#ifdef LINUX_ALPINE` for this case.
> * Abandon the possibility of dynamically supporting Linux & Alpine.

I (really) like the possibility of dropping a couple of dll's, but I was wondering what it'd buy us on the long run. I don't know if I am lacking experience or prior art here, but I don't see much of an issue creating a separate one. Or if it warrants this complexity

> * Define a mechanism for the hosting layer to inform the elf-reader the type of dump.

This one is nice (no guessing) but rarely a good option in my opinion. The DAC will need to now gain more information of the environment it runs in, and so will createdump and such.

Overall, I would say I don't mind a heuristic that much. I just don't know if:

+ 176 is correct.
+ An idea of how likely it is to SEGV trying of reading memory (definitely a LOT less now that there's a segment constrained read).
+ What does this buy us in the end in exchange for the risk.